### ISSUE TITLE
ENH add mitigation algorithm "Mechanisms for Fair Classification" by Zafar et al.

### ISSUE THREAD

**Issue by hildeweerts** [fairlearn/fairlearn#1025]
The paper [Fairness Constraints: Mechanisms for Fair Classification](http://proceedings.mlr.press/v54/zafar17a/zafar17a.pdf) by Zafar et al. introduces a constrained learning algorithm for demographic parity. The goal of this task is to implement the **logistic regression** algorithm described in this paper in Fairlearn.

* An existing Python implementation of the authors can be found [here](https://github.com/mbilalzafar/fair-classification/tree/master/disparate_impact).
* The paper describes two different formulations: (1) maximizing accuracy under fairness constraints, and (2) maximizing fairness under accuracy constraints. We can choose one of them (I think (1) is most natural) or implement both.
* I would suggest to design the API similar to [`sklearn.linear_model.LogisticRegression`](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LogisticRegression.html).

Completing this item requires:
- [ ] technique code in `fairlearn.linear_model` (this is my suggestion following scikit-learn, but the specific naming is up for discussion).
- [ ] unit tests in `test.unit.linear_model`
- [ ] descriptive API reference (directly in the docstring)
- [ ] a short user guide in `docs.user_guide.mitigation.rst`

A fully fledged example notebook is not required.

To claim this task please respond below. Of course, you can also ask questions!

**Comment by rensoostenbach** [comment]
I will start on this issue next week :). While working on this, should I in some way keep in mind that this is closely related to #1027, or is it just meant as extra information?

**Comment by hildeweerts** [comment]
Awesome @rensoostenbach!  In case somebody else starts working on #1027 it would be good to coordinate, but for now it is just additional info.

**Comment by rensoostenbach** [comment]
My plan to approach this is as follows:
I will create a `FairLogisticRegression` class in `fairlearn.linear_model._fair_logistic.py`. This class inherits from the sklearn `LogisticRegression` and will be changed in the places where adaptations need to be made.
These adaptations are as follows:

- Adding a `constraints` parameter to the `optimize.minimize` function ([here in sklearn](https://github.com/scikit-learn/scikit-learn/blob/7e1e6d09bcc2eaeba98f7e737aac2ac782f0e5f1/sklearn/linear_model/_logistic.py#L806-L813), as is done in the [code of the paper](https://github.com/mbilalzafar/fair-classification/blob/45e85baad209134bbe2697c1a884dd1e2c2d9786/fair_classification/utils.py#L57-L71))
- Making sure that the `FairLogisticRegression` can handle sensitive attributes. Would this be as easy as dropping them from `X` such that the model does not use them for training, or is there something else I should be aware of?

The plan is to start with only the `lbfgs` solver, since it uses the same `optimize.minimize` function in sklearn as well as the code of the paper. Also, it is the default solver for `LogisticRegression`. 

I would like to get some feedback on this approach, and perhaps some things I should take caution of when inheriting from `LogisticRegression`, etc. Thanks!

**Comment by adrinjalali** [comment]
sklearn has a new GLM framework with a few models implemented under it, and it'd be nice to see if you could extend those instead of implementing from scratch. An example is `TweedieRegressor`: https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.TweedieRegressor.html

**Comment by rensoostenbach** [comment]
> sklearn has a new GLM framework with a few models implemented under it, and it'd be nice to see if you could extend those instead of implementing from scratch. An example is `TweedieRegressor`: https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.TweedieRegressor.html

This seems to be relatively doable! From the [source code](https://github.com/scikit-learn/scikit-learn/blob/7e1e6d09bcc2eaeba98f7e737aac2ac782f0e5f1/sklearn/linear_model/_glm/glm.py#L310-L322), the same `optimize.minimize` function is used, and the `TweedieRegressor` only uses the `lbfgs` solver, making it quite easy to extend it to.
